### ISSUE TITLE
Caching with Redis on a Parse Server

### ISSUE THREAD

**Issue by Samigos** [parse-community/parse-server#4932]
I'm on a 2.6.3 Parse Server and I need to cache the results of queries, to speed things up!

I understand that Parse Server offers a Redis adapter. What exactly do I have to do, in order to start using Redis? Are there any modules I should install? Anything I should import or configure?

I found this on Parse's documentation: `Those cache adapters can be cleaned at anytime internally, you should not use them to cache data and you should let parse-server manage their data lifecycle.`

What does it mean by saying `you should not use them to cache data and you should let parse-server manage their data lifecycle.`? Should I not use the adapter?

**Comment by flovilmart** [comment]
The caches are internal, and you should not need to configure them unless you know you need a shared cache instead of the in memory cache AND you understand the problem it solves.

> I'm on a 2.6.3 Parse Server and I need to cache the results of queries, to speed things up!

This is not the purpose of the redis cache. The Redis cache, that you can optionally configure can be used to share data across multiple servers like session tokens / users etc... In no case, this allows what you're attempting to do.

Also consider asking the question on stackoverflow / stack exchange as the caching mechanism you're looking for can be achieved multiple ways.

**Comment by Samigos** [comment]
@flovilmart So I shouldn't use Redis at all? Could you suggest something else then? Something that would play well along with parse server...

**Comment by flovilmart** [comment]
What I am saying is that if you want to replace the internal cache management from memory to redis you can use the RedisCacheAdapter.

Otherwise, if you want to implement caching, you *can* use redis, as you can use *memcached*. The RedisCacheAdapter is just a simple cache object that implements the CacheAdapter interface. in theory you can use it as long as your problem can be solved by the interface provided by the cache (https://github.com/parse-community/parse-server/blob/master/src/Adapters/Cache/CacheAdapter.js)

BUT, you have to write your own caching mechanisms, to check for stale data, invalidate, store etc...

**Comment by oallouch** [comment]
I guess he'd like a snippet on how to serialize/deserialize Parse Objects.

**Comment by flovilmart** [comment]
SO you can provide him with that, I'm not sure how he want to achieve his caching. Caching is such a wide subject there are many implementations possible.

**Comment by Samigos** [comment]
Why not? Isn't that why we use caching, to minimize latency?

**Comment by flovilmart** [comment]
> Isn't that one of the main reasons why we use caching, to minimize latency?

Caching minimized latency because it reduces the contention on the DB. A properly indexed database should be able to respond quickly, the full round trip for common requests should be around 25/30ms, if you poke from a cache, you'll get to 10/15ms (as you still have the latency from the server to the cache server).

What are your average server response times? From the time the request comes in to the response is written?

Also serialization in JSON of large objects blocks the node process, when you use a cache, you need to serialize those object. So serializing / deserializing thousand's of objects for the cache may make your server less responsive see: https://github.com/parse-community/parse-server/pull/4135

To sum up, using a cache may help you but your mileage will vary, depending on your expertise in getting out all the performance from a non cached system.
